logstash-scheduler/src/main/java/org/apache/mesos/logstash/scheduler/OfferStrategy.java

```diff
@@ -62,76 +64,71 @@ public OfferResult evaluate(ClusterState clusterState, Protos.Offer offer) {
      * Offer result
      */
     public static class OfferResult {
-        final boolean acceptable;
-        final Optional<String> reason;
+        final List<String> complaints;
 
-        private OfferResult(boolean acceptable, Optional<String> reason) {
-            this.acceptable = acceptable;
-            this.reason = reason;
+        private OfferResult(List<String> complaints) {
+            this.complaints = complaints;
+        }
+
+        public boolean acceptable() {
+            return complaints.isEmpty();
         }
 
         public static OfferResult accept() {
-            return new OfferResult(true, Optional.<String>empty());
+            return new OfferResult(Arrays.asList());
         }
 
-        public static OfferResult decline(String reason) {
-            return new OfferResult(false, Optional.of(reason));
+        public static OfferResult decline(List<String> complaints) {
+            return new OfferResult(complaints);
         }
     }
 
-    private boolean isHostAlreadyRunningTask(ClusterState clusterState, Protos.Offer offer) {
-        return clusterState.getTaskList().stream().anyMatch(taskInfo -> taskInfo.getSlaveId().equals(offer.getSlaveId()));
+    private Stream<String> hostRule(ClusterState clusterState, Protos.Offer offer) {
+        return clusterState
+                .getTaskList()
+                .stream()
+                .filter(taskInfo -> taskInfo.getSlaveId().equals(offer.getSlaveId()))
+                .map(taskInfo -> "host " + taskInfo.getSlaveId().getValue() + " is already running task " + taskInfo.getTaskId().getValue());
     }
 
-    private boolean hasEnoughOfResourceType(List<Protos.Resource> resources, String resourceName, double minSize) {
-        for (Protos.Resource resource : resources) {
-            if (resourceName.equals(resource.getName())) {
-                return resource.getScalar().getValue() >= minSize;
-            }
+    private Stream<String> complaintsForResourceType(List<Protos.Resource> resources, String resourceName, double minSize) {
+        double totalSize = resources.stream().filter(resource -> resource.getName().equals(resourceName)).collect(Collectors.summingDouble(resource -> resource.getScalar().getValue()));
+        if (totalSize < minSize) {
+            return Arrays.asList("required minimum " + minSize + " " + resourceName + " but offer only has " + totalSize + " in total").stream();
```

Comment:
using `java.util.Collections.singletonList()` is a lot safer, because it's also immutable